descs: refactor GetAll* methods

postamar opened the pull request:
This commit refactors the descs.Collection's GetAll* methods, which
return descriptors and other catalog metadata in bulk, to share a common
access path in which the descriptor's layers are properly aggregated.
This is non-trivial as there are numerous edge cases:
  - virtual schemas and objects exist for all databases and their
    descriptors don't reference the parent database at all,
  - descriptorless public schemas and temporary schemas only exist as
    namespace entries,
  - functions don't have namespace entries,
  - and so forth.

These new methods are delegated to by the existing methods which are now
deprecated.

Informs #64089.

Release note: None

Commit:
nstree: Catalog method cleanup

This commit renames existing methods and adds a few missing pieces for
the sake of consistency.

Release note: None

Commit:
catalog: add Visit to VirtualSchemas

This commit adds a new Visit method to the VirtualSchemas interface
which allows traversing the entirety of the virtual catalog.

Release note: None

Commit:
descs: refactor GetAll* methods

This commit refactors the descs.Collection's GetAll* methods, which
return descriptors and other catalog metadata in bulk, to share a common
access path in which the descriptor's layers are properly aggregated.
This is non-trivial as there are numerous edge cases:
  - virtual schemas and objects exist for all databases and their
    descriptors don't reference the parent database at all,
  - descriptorless public schemas and temporary schemas only exist as
    namespace entries,
  - functions don't have namespace entries,
  - and so forth.

These new methods are delegated to by the existing methods which are now
deprecated.

Informs #64089.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/93543)
<!-- Reviewable:end -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 32 of 32 files at r1, 2 of 2 files at r2, 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93543#-:-NJC8Oe06Gfae0b7tXgD:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @postamar)

___
*[`pkg/sql/catalog/descs/collection.go` line 861 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/93543#-NJC7cDE9KEnua-wDk1Y:-NJC7cDE9KEnua-wDk1Z:bsu5kql) ([raw file](https://github.com/cockroachdb/cockroach/blob/767e07f44ca76dc7261edc5239d1abc0c6c2dc0f/pkg/sql/catalog/descs/collection.go#L861)):*
> ```Go
> 	ctx context.Context, txn *kv.Txn, stored nstree.Catalog, seedIDs catalog.DescriptorIDSet,
> ) (ret nstree.MutableCatalog, _ error) {
> 	descIDs := catalog.MakeDescriptorIDSet(seedIDs.Ordered()...)
> ```

nit: it'd be nice to not need to make this slice, but it's a nit
___
*[`pkg/sql/catalog/tabledesc/table_desc.go` line 78 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/93543#-NJC78wv44AygXRX6y4A:-NJC78wv44AygXRX6y4B:bf210lc) ([raw file](https://github.com/cockroachdb/cockroach/blob/767e07f44ca76dc7261edc5239d1abc0c6c2dc0f/pkg/sql/catalog/tabledesc/table_desc.go#L78)):*
> ```Go
> // SkipNamespace implements the descriptor interface.
> func (desc *wrapper) SkipNamespace() bool {
> 	return desc.IsVirtualTable()
> ```

This could use some commentary


<!-- Sent from Reviewable.io -->

postamar commented:
Thanks for the review. I've addressed your comments. Notice also that the rttanalysis benchmark regresses with this PR; I'm well aware & this will be addressed shortly. There's some duplication of work due to some uncalled-for use of the `catkv.Direct` interface in `pg_catalog.go`, it's quite straightforward.

postamar commented:
bors r+
